creator-node/src/snapbackSM.js

```diff
@@ -47,63 +58,308 @@ class SnapbackSM {
     )
   }
 
+  // Helper function to retrieve all config based
+  async getSPInfo () {
+    const spID = config.get('spID')
+    const endpoint = this.endpoint
+    const delegateOwnerWallet = config.get('delegateOwnerWallet')
+    const delegatePrivateKey = config.get('delegatePrivateKey')
+    return {
+      spID,
+      endpoint,
+      delegateOwnerWallet,
+      delegatePrivateKey
+    }
+  }
+
+  // Query eth-contracts chain for endpoint to ID info
+  async recoverSpID () {
+    if (config.get('spID') !== 0) {
+      this.log(`Known spID=${config.get('spID')}`)
+      return
+    }
+
+    const recoveredSpID = await this.audiusLibs.ethContracts.ServiceProviderFactoryClient.getServiceProviderIdFromEndpoint(
+      this.endpoint
+    )
+    this.log(`Recovered ${recoveredSpID} for ${this.endpoint}`)
+    config.set('spID', recoveredSpID)
+  }
+
+  // Retrieve users with this node as primary
+  async getNodePrimaryUsers () {
+    const currentlySelectedDiscProv = this.audiusLibs.discoveryProvider.discoveryProviderEndpoint
+    let requestParams = {
+      method: 'get',
+      baseURL: currentlySelectedDiscProv,
+      url: `users/creator_node`,
+      params: {
+        creator_node_endpoint: this.endpoint
+      }
+    }
+    let resp = await axios(requestParams)
+    this.log(`Discovery provider: ${currentlySelectedDiscProv}`)
+    return resp.data.data
+  }
+
   /*
-    Main state machine processing function
-    TODO: Update job arguments
+    Retrieve the current clock value on this node for the provided user wallet
   */
+  async getUserPrimaryClockValue (wallet) {
+    let walletPublicKey = wallet.toLowerCase()
+    const cnodeUser = await models.CNodeUser.findOne({
+      where: { walletPublicKey }
+    })
+    const clockValue = (cnodeUser) ? cnodeUser.dataValues.clock : -1
+    return clockValue
+  }
+
+  // Enqueue a sync request to a particular secondary
+  async issueSecondarySync (userWallet, secondaryEndpoint, primaryEndpoint) {
+    let syncRequestParameters = {
+      baseURL: secondaryEndpoint,
+      url: '/sync',
+      method: 'post',
+      data: {
+        wallet: [userWallet],
+        creator_node_endpoint: primaryEndpoint,
+        state_machine: true // state machine specific flag
```

Comment:
this flag doesn't do anything?

Reply:
true

Reply:
you know i think it is worth us logging this out on the other side just so we have an 'initiated by state_machine' from sync side - i think that was the original intent behind this